options/options.go

```diff
@@ -0,0 +1,118 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0'
+ */
+
+package options
+
+import (
+	"encoding/json"
+
+	"github.com/canonical/edgex-snap-hooks/v2/log"
+	"github.com/canonical/edgex-snap-hooks/v2/snapctl"
+)
+
+type SnapOptions struct {
+	Apps   map[string]map[string]map[string]interface{} `json:"apps"`
+	Config map[string]interface{}                       `json:"config"`
+}
+
+func getConfigMap(config map[string]interface{}) (map[string]string, error) {
+	result := make(map[string]string)
+
+	for env, value := range config {
+		if err := flattenConfigJSON("", env, value, result); err != nil {
+			return nil, err
+		}
+	}
+	return result, nil
+}
+
+func ProcessOptions(services []string) error {
+	var options SnapOptions
+
+	//Schema (4): 	config.<my.env.var> -> setting env variable for all apps (e.g. DEBUG=true, SERVICE_SERVERBINDADDRESS=0.0.0.0)
+
+	jsonString, err := snapctl.Get("config").Document().Run()
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(jsonString), &options)
+	if err != nil {
+		return err
+	}
+	configuration, err := getConfigMap(options.Config)
+	if err != nil {
+		return err
+	}
+
+	for _, service := range services {
+		overrides := getEnvVarFile(service)
+		for env, value := range configuration {
+			overrides.setEnvVariable(env, value)
+		}
+		overrides.writeEnvFile(false)
+	}
+
+	//Schema (1): apps.<app>.config.<my.env.var> -> setting env var MY_ENV_VAR for an app
+
+	// get the 'apps' json structure
+	jsonString, err = snapctl.Get("apps").Document().Run()
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(jsonString), &options)
+	if err != nil {
+		return err
+	}
+
+	// iterate through the known services in this snap
+	for _, service := range services {
+		log.Infof("Processing service:%s", service)
+
+		// get the configuration specified for each service
+		// and create the environment override file
+		appConfig := options.Apps[service]
+		log.Infof("Processing appConfig:%v", appConfig)
+		if appConfig != nil {
+			config := appConfig["config"]
+			log.Infof("Processing config:%v", config)
+			if config != nil {
+				configuration, err := getConfigMap(config)
+
+				log.Infof("Processing configuration:%v", configuration)
+				if err != nil {
+					return err
+				}
+				overrides := getEnvVarFile(service)
+
+				log.Infof("Processing overrides:%v", overrides)
+				for env, value := range configuration {
+					log.Infof("Processing overrides setEnvVariable:%v %v", env, value)
+					overrides.setEnvVariable(env, value)
+				}
+				overrides.writeEnvFile(true)
+			}
+		}
+	}
+
+	// apps.<app>.<other-option.child> -> setting another option for CLI executation or CLI arg override
+
+	// apps.<app>.auto-start (boolean) -> turn auto start on/off by seting to true/false
+
+	return nil
+
+}
```

Comment:
I think this function is designed to do too much inline.

I'd separate the different groups of options into separate functions and call them. Example:

ProcessOptions:
- processConfigOptions
  - processGlobalConfigOptions
  - processAppConfigOptions
- processAutostartOptions
- processProxyOptions -> didn't we decide to move this out of the library?

Reply:
split this up